hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java

```diff
@@ -67,30 +72,64 @@ public ClientPartitionServiceImpl(HazelcastClientInstanceImpl client) {
     }
 
+    public void listenPartitionTable(Connection ownerConnection) throws Exception {
+        //when we connect to cluster back we need to reset partition state version
+        lastPartitionStateVersion = -1;
+        if (((ClientConnection) ownerConnection).getConnectedServerVersion() >= BuildInfo.calculateVersion("3.9")) {
+            //Servers after 3.9 supports listeners
+            ClientMessage clientMessage = ClientAddPartitionListenerCodec.encodeRequest();
+            ClientInvocation invocation = new ClientInvocation(client, clientMessage, ownerConnection);
+            invocation.setEventHandler(this);
+            invocation.invokeUrgent().get();
+        }
+    }
+
     public void refreshPartitions() {
-        ClientExecutionServiceImpl executionService = (ClientExecutionServiceImpl) client.getClientExecutionService();
         try {
             // use internal execution service for all partition refresh process (do not use the user executor thread)
-            executionService.execute(new RefreshTask());
+            clientExecutionService.execute(new RefreshTask());
         } catch (RejectedExecutionException ignored) {
             EmptyStatement.ignore(ignored);
         }
     }
 
-    private void getPartitionsBlocking() {
-        while (!getPartitions() && client.getConnectionManager().isAlive()) {
+    @Override
+    public void handle(Collection<Map.Entry<Address, List<Integer>>> collection, int partitionStateVersion) {
+        processPartitionResponse(collection, partitionStateVersion, true);
+    }
+
+    @Override
+    public void beforeListenerRegister() {
+
+    }
+
+    @Override
+    public void onListenerRegister() {
+
+    }
+
+    private void waitForPartitionsFetchedOnce() {
+        while (partitionCount == 0 && client.getConnectionManager().isAlive()) {
             if (isClusterFormedByOnlyLiteMembers()) {
                 throw new NoDataMemberInClusterException(
                         "Partitions can't be assigned since all nodes in the cluster are lite members");
             }
+            ClientMessage requestMessage = ClientGetPartitionsCodec.encodeRequest();
+            ClientInvocationFuture future = new ClientInvocation(client, requestMessage).invokeUrgent();
             try {
-                Thread.sleep(PARTITION_WAIT_TIME);
-            } catch (InterruptedException e) {
-                e.printStackTrace();
+                ClientMessage responseMessage = future.get();
+                ClientGetPartitionsCodec.ResponseParameters response =
+                        ClientGetPartitionsCodec.decodeResponse(responseMessage);
+                processPartitionResponse(response.partitions,
+                        response.partitionStateVersion, response.partitionStateVersionExist);
+            } catch (Exception e) {
+                if (client.getLifecycleService().isRunning()) {
```

Comment:
without no sleep and with current codec this causes busy loop. we need to look if we can change the codec to be retryable message. We hope that the exception will mostly be IOException or the InstanceNotActiveException, and hence hopefully will go to retry mechanism which does the sleep to slow down the busy loop.